### DISCUSSION TITLE
GraphQL unable to resolve Global Id Ruleset

### DISCUSSION THREAD

**Discussion by kmason-extron** [community/community#60620] [Apps, API and Webhooks]
### Select Topic Area

Bug

### Body

I want to be able to add bypass actors via their Global Ids in `updateRepositoryRuleset` or `createRepositoryRuleset` but get the following message

> `"Could not resolve to a node with the global id of 'T_xxxxxxxxxxxxxxxxx'."` I obfuscated our Id.

I have no problems with the other recent changes for using org admins, repository roles etc only when specifying `actorIds`

If I update them in the UI and query the ruleset then the returned info is correct and shows them but using graphql to set them does not.

### Mutation

```
mutation UpdateRepoRule(
    $ruleId: ID!, 
    $name: String!, 
    $enforcement: RuleEnforcement!, 
    $target: RepositoryRulesetTarget!, 
    $include: [String!]!, 
    $type: RepositoryRuleType!, 
    $parameters: RuleParametersInput!,
    $bypass: [RepositoryRulesetBypassActorInput!]
  ) {
  updateRepositoryRuleset(
    input: {
        repositoryRulesetId: $ruleId, 
        name: $name, 
        enforcement: $enforcement, 
        target: $target,
        bypassActors: $bypass,
        conditions: {
            refName: {
                exclude: [],
                include: $include
            }
        },
        rules: [
            {
                type: $type,
                parameters: $parameters
            }
        ]
    }
  ) {
    ruleset {
      id,
      databaseId 
    }
  }
}
``` 

### Variables

```
{
  "ruleId": "RRS_xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx",
  "target": "BRANCH",
  "include": [
    "~DEFAULT_BRANCH",
    "refs/heads/hotfix.*",
    "refs/heads/dvt.*"
  ],
  "enforcement": "ACTIVE",
  "bypass": [
    {
      "bypassMode": "ALWAYS",
      "repositoryRoleDatabaseId": 5
    },
    {
      "bypassMode": "ALWAYS",
      "organizationAdmin": true
    },
    {
      "bypassMode": "ALWAYS",
      "actorId": "T_xxxxxxxxxxxxxxxxxxx"
    }
  ],
  "type": "REQUIRED_STATUS_CHECKS",
  "name": "PR Status Checks",
  "parameters": {
    "requiredStatusChecks": {
      "strictRequiredStatusChecksPolicy": true,
      "requiredStatusChecks": []
    }
  }
}
``` 

### Full Response 

```
{
  "data": {
    "updateRepositoryRuleset": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "updateRepositoryRuleset"
      ],
      "locations": [
        {
          "line": 11,
          "column": 3
        }
      ],
      "message": "Could not resolve to a node with the global id of 'T_xxxxxxxxxxxxxxxxx'."
    }
  ]
}
``` 

### Query after manual update in UI

```
{
  "data": {
    "repository": {
      "rulesets": {
        "nodes": [
          {
            "name": "PR Status Checks",
            "bypassActors": {
              "nodes": [
                {
                  "repositoryRoleName": "admin",
                  "repositoryRoleDatabaseId": 5,
                  "organizationAdmin": false,
                  "actor": null,
                  "bypassMode": "ALWAYS"
                },
                {
                  "repositoryRoleName": null,
                  "repositoryRoleDatabaseId": null,
                  "organizationAdmin": false,
                  "actor": {
                    "id": "T_xxxxxxxxxxxxxxxxxx",
                    "name": "Control Software Engineering Platform"
                  },
                  "bypassMode": "ALWAYS"
                },
                {
                  "repositoryRoleName": null,
                  "repositoryRoleDatabaseId": null,
                  "organizationAdmin": false,
                  "actor": {
                    "id": "T_xxxxxxxxxxxxxxxxxx",
                    "name": "Control Software Engineering Leads"
                  },
                  "bypassMode": "ALWAYS"
                },
                {
                  "repositoryRoleName": null,
                  "repositoryRoleDatabaseId": null,
                  "organizationAdmin": true,
                  "actor": null,
                  "bypassMode": "ALWAYS"
                }
              ]
            }
          }
        ]
      }
    }
  }
}
```

**Comment by kmason-extron** [comment]
It does appear that the REST endpoint works (substitution of databaseId for the string in graph)

**Comment by patrick-knight** [comment]
Thanks for the info @kmason-extron just to confirm we're looking into this.

**Reply by kmason-extron** [reply]
Thanks. Let me know if you need more testing from me. So far only the rulesets I've found has this issue. Everywhere else is using the int or string versions just fine.

**Reply by kmason-extron** [reply]
@patrick-knight with the GA notice has this been resolved?

**Reply by patrick-knight** [reply]
Hey @kmason-extron sorry for the delayed response I was out of the office for a bit. 

Looking back at notes with the team we're having trouble replicating the same behavior. Would you mind opening a support case so you can share more details about the org, repo, actor, etc. so we can investigate more in depth? 

Thanks!

**Comment by rfossicav** [comment]
Greetings Team, any updates on this?  I'm having a similar issue with `createRepositoryRuleset`, but the issue seems to be against the required `sourceId` field. I've tried with the legacy id and with the `X-Github-Next-Global-ID` id the curl spit out.   Or maybe Maybe I'm not understanding the syntax / usage here? 

the call: 
```
mutation {
  createRepositoryRuleset(input: 
    {
      sourceId: "U_*********"
      name: "branch rule test 001"
      enforcement: ACTIVE
      target: BRANCH
      conditions: {
        repositoryName: {
          include: ["real-repo-name-here"]
          exclude: []
        }
      }
      rules: {
      type:CREATION
      parameters: {
        branchNamePattern: {
          pattern: "*001test*"
          operator: "equals"
        }
      }
    }
      bypassActors: [
        {
        bypassMode: ALWAYS
        organizationAdmin: true
      	}
      ]
    }) 
    {
    clientMutationId
    } 
  }
```

The response to my mutation: 
```
{
  "data": {
    "createRepositoryRuleset": null
  },
  "errors": [
    {
      "type": "NOT_FOUND",
      "path": [
        "createRepositoryRuleset"
      ],
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "message": "Could not resolve to RuleSource node with the global id of 'U_***********'."
    }
  ]
}
```

**Reply by jschnapper** [reply]
Hey @rfossicav, I think there may be some confusion on what id to provide as an input. The `sourceId` accepts the global node id for the source (a repository or organization) of a ruleset. More info on global node ids here: https://docs.github.com/en/graphql/guides/using-global-node-ids

It looks like you're making an organization level ruleset, so you'll want to get the global node id of the organization and provide it as the input.

Also, a few other notes:
- The `rules` field takes an array of rule inputs, which you can see from the `CreateRepositoryRulesetInput` documentation: https://docs.github.com/en/graphql/reference/input-objects#createrepositoryrulesetinput
- The `CREATION` rule does not have a parameters field, which can be seen from its omission from the `RuleParametersInput`: https://docs.github.com/en/graphql/reference/input-objects#ruleparametersinput. However, I think the it's most clear to see what rules have parameters by looking at the REST API docs: https://docs.github.com/en/rest/repos/rules?apiVersion=2022-11-28#create-a-repository-ruleset
![creation rule](https://github.com/community/community/assets/15844781/a67d0ec4-be30-4df5-9d6a-c009df125dc3)
- Instead of adding `branchNamePattern` to the `CREATION` rule, you likely want to add a `refName` field to your conditions: https://docs.github.com/en/graphql/reference/input-objects#repositoryruleconditionsinput.

For example

```graphql
      conditions: {
        repositoryName: {
          include: ["real-repo-name-here"]
          exclude: []
        }
        refName: {
          include: ["*001test*"]
        }
      }
      rules: [
        {
            type: CREATION
        }
      ]
    }

```

**Reply by rfossicav** [reply]
@jschnapper Thanks so much! This helped me a lot... It helped me wrap my brain around the parameters and where they actually belonged.

My final call for posterity:
```
mutation {
  createRepositoryRuleset(input: 
    {
      sourceId: "<<repo_node_id_here>>" # must use repo node_id
      name: "Restrict Honey Branch Usage" # your ruleset name
      enforcement: ACTIVE
      target: BRANCH
      conditions: {
        refName: {
          include: ["refs/heads/*honey*"] # keep refs/heads/ portion
          exclude: []
        }
      }
      rules: [
        {
          type:CREATION 
        },
        {
          type:NON_FAST_FORWARD
        },
        {
          type:DELETION
        },
        {
          type:PULL_REQUEST 
            parameters: {
                pullRequest: {
                    requiredApprovingReviewCount: 2
                    requireCodeOwnerReview: true
                    dismissStaleReviewsOnPush: false
                    requireLastPushApproval: false
                    requiredReviewThreadResolution: false
                }
            }
        }
      ]
  	}) 
  {
    clientMutationId
  } 
}
```

Next, I will work on improving my calls and using variables.  Thank you again!

**Comment by jhon615-ctr** [comment]
This users are really using the key and remote system for some hacking and humanization also doing like feeds and conversation, sexual content with they are characteristics like wife mommy really almost 4 month travel with them it's horrible. They are characteristics and frame works for me very terrible for locally and my colleagues and my family. My devices are currently cracked. I met all most 500 oman riyal loses. And My sleepless nights please block all users and their all transactions and all programs stop and block and please respond for my loses really hard devil's are working their .and my Gmail account also they are binding 

![Screenshot_20250612_070927](https://github.com/user-attachments/assets/b08c8f87-9422-4df8-9453-c37e856220b4)
![Screenshot_20250612_070956](https://github.com/user-attachments/assets/0bbb882f-80fc-4aa2-97c0-1aaa521692b6)
![Screenshot_20250612_070850](https://github.com/user-attachments/assets/84334903-2bce-4f0a-a57e-f6048a8594ab)

**Comment by AndreasPetersen** [comment]
I had the same issue. The issue is that the documentation is incorrect.

The documentation for [RepositoryRulesetBypassActorInput](https://docs.github.com/en/graphql/reference/input-objects#repositoryrulesetbypassactorinput) states:

> actorId (ID): For Team and Integration bypasses, the Team or Integration ID.

However, this is incorrect. What you need to specify is the node ID, not the ID. Using the [Get an app](https://docs.github.com/en/rest/apps/apps?apiVersion=2022-11-28#get-an-app) REST endpoint, you get a response that contains an `id` and a `node_id`. Based on the documentation for `RepositoryRulesetBypassActorInput`, you would think you should provide the `id`, but you actually need to specify the `node_id`.
